### PR TITLE
added neurodamus-asciitoh as requested in BBPBGLIB-512

### PR DESCRIPTION
**Pull request by st4rl3ss**
Simple addition of mod and hoc files necessary to run `convert.sh` to convert files from the ascii format to HDF5. 

The files were taken from the `neurodamus-core` repo and moved to a different specific repo: `https://bbpcode.epfl.ch/code/#/admin/projects/sim/utils/asciitoh5` 

The rest should be a relatively simple recipe.

### CODE REVIEW

File: var/spack/repos/builtin/packages/neurodamus-asciitoh/package.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        2. shared libs passed with -L -l and RPATH flags
+        Attention: This func doesnt handle recursive deps of static libs.
+        """
+        spec = self.spec[lib_name]
+        if spec.satisfies('+shared'):  # Prefer shared if both exist
+            return "%s %s" % (spec.libs.rpath_flags, spec.libs.ld_flags)
+        return spec.libs.joined()
+
+    def install(self, spec, prefix):
+        shutil.copytree('hoc', prefix.hoc)
```
Comment (iomaganaris) [inline var/spack/repos/builtin/packages/neurodamus-asciitoh/package.py]: I was thinking after seeing the install directory of `neurodamus-neocortex` for example that `hoc`, `mod` and `x86_64` directories would be better to be moved in a `lib` folder.
```suggestion
os.mkdir(prefix.lib)
shutil.copytree('hoc', prefix.lib.hoc)
...
```

**Review by iomaganaris — COMMENTED** [review]
Overall LGTM. I have a small suggestion that came up to my mind only

File: var/spack/repos/builtin/packages/neurodamus-asciitoh/package.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
+# Copyright 2013-2018 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+
+from spack import *
+import shutil
+import os
+
+class NeurodamusAsciitoh(Package):
```
Comment (matz-e) [inline var/spack/repos/builtin/packages/neurodamus-asciitoh/package.py]: How about `ascii2h`?

Comment (matz-e) [reply var/spack/repos/builtin/packages/neurodamus-asciitoh/package.py]: Also note that the URL has an additional 5 in there. Is this ASCII to headers?

File: var/spack/repos/builtin/packages/neurodamus-asciitoh/package.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        2. shared libs passed with -L -l and RPATH flags
+        Attention: This func doesnt handle recursive deps of static libs.
+        """
+        spec = self.spec[lib_name]
+        if spec.satisfies('+shared'):  # Prefer shared if both exist
+            return "%s %s" % (spec.libs.rpath_flags, spec.libs.ld_flags)
+        return spec.libs.joined()
+
+    def install(self, spec, prefix):
+os.mkdir(prefix.lib)
```
Comment (matz-e) [inline var/spack/repos/builtin/packages/neurodamus-asciitoh/package.py]: Github gives me a funky indent here...

**Review by matz-e — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/neurodamus-asciitoh/package.py (hunk 2 above)
Comment (st4rl3ss) [reply var/spack/repos/builtin/packages/neurodamus-asciitoh/package.py]: `ascii2h` sounds nicer.
I think I had issues when a spack package ended with a number. It was meant to be `asciitoh5` but I would get strange errors I don't exactly recall when I created it. Can you confirm if there's any issues with creating a package whose name ends with a number?

File: var/spack/repos/builtin/packages/neurodamus-asciitoh/package.py (hunk 3 above)
Comment (st4rl3ss) [reply var/spack/repos/builtin/packages/neurodamus-asciitoh/package.py]: Yeah, I'm fixing it manually right now.

File: var/spack/repos/builtin/packages/neurodamus-asciitoh/package.py (hunk 2 above)
Comment (st4rl3ss) [reply var/spack/repos/builtin/packages/neurodamus-asciitoh/package.py]: Also, it's meant to be `asciitohdf5`.

Comment (matz-e) [reply var/spack/repos/builtin/packages/neurodamus-asciitoh/package.py]: I don't see why there should be issues. Just ran across a package called `py-rpy2` [sic].

`asciitoh5` would be in line with the repository URL.

Comment (pramodk) [reply var/spack/repos/builtin/packages/neurodamus-asciitoh/package.py]: > asciitoh5 would be in line with the repository URL

+2

> I think I had issues when a spack package ended with a number. It was meant to be asciitoh5 but I would get strange errors I don't exactly recall when I created it.

I don't think this is related. If you an error, could you post?

File: var/spack/repos/builtin/packages/neurodamus-asciitoh/package.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+class NeurodamusAsciitoh(Package):
+    """Neurodamus Library necessary to convert from ASCII to H5"""
+   
+    homepage = "ssh://bbpcode.epfl.ch/sim/utils/asciitoh5"
+    git      = "ssh://bbpcode.epfl.ch/sim/utils/asciitoh5"
+
+    version('develop', git=git, branch='master')
+    version('0.1', git=git, tag='0.1')
+
+    depends_on('neuron~binary+python~mpi')
```
Comment (pramodk) [inline var/spack/repos/builtin/packages/neurodamus-asciitoh/package.py]: why `~binary+python`? Is python required for the scripts?

File: var/spack/repos/builtin/packages/neurodamus-asciitoh/package.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+            return "%s %s" % (spec.libs.rpath_flags, spec.libs.ld_flags)
+        return spec.libs.joined()
+
+    def install(self, spec, prefix):
+os.mkdir(prefix.lib)
+shutil.copytree('hoc', prefix.lib.hoc)
+...
+        shutil.copytree('mod', prefix.mod)
+	with working_dir(prefix):
+	    link_flag = '-Wl,-rpath,' + self._get_link_flags('hdf5')
```
Comment (pramodk) [inline var/spack/repos/builtin/packages/neurodamus-asciitoh/package.py]: I wonder if `_get_link_flags` is required or can't be simplified here.

File: var/spack/repos/builtin/packages/neurodamus-asciitoh/package.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+	    include_flag = ' -I%s' % (spec['hdf5'].prefix.include)
+            which('nrnivmodl')('-incflags', include_flag, '-loadflags', link_flag, 'mod')
+	    bindir = os.path.basename(self.neuron_archdir)
+            special = join_path(bindir, 'special')
+            os.mkdir(prefix.bin)
+	    shutil.copy(special, prefix.bin)
+            for f in find('hoc/neuronHDF5', 'convert.sh'):
+                shutil.copy(f, prefix.bin)
```
Comment (pramodk) [inline var/spack/repos/builtin/packages/neurodamus-asciitoh/package.py]: `hoc/neuronHDF5` shouldn't be in `bin` directory?

**Review by pramodk — CHANGES_REQUESTED** [review]
cc : @ferdonline

File: var/spack/repos/builtin/packages/neurodamus-asciitoh/package.py (hunk 4 above)
Comment (st4rl3ss) [reply var/spack/repos/builtin/packages/neurodamus-asciitoh/package.py]: I just used the same dependency of `neurodamus`. If you think `python` is useless I can remove it.

File: var/spack/repos/builtin/packages/neurodamus-asciitoh/package.py (hunk 2 above)
Comment (st4rl3ss) [reply var/spack/repos/builtin/packages/neurodamus-asciitoh/package.py]: I will recreate it as `asciitoh5` then.

File: var/spack/repos/builtin/packages/neurodamus-asciitoh/package.py (hunk 5 above)
Comment (st4rl3ss) [reply var/spack/repos/builtin/packages/neurodamus-asciitoh/package.py]: to keep things clean I thought of using this function that is used elsewhere in `spack`.

File: var/spack/repos/builtin/packages/neurodamus-asciitoh/package.py (hunk 6 above)
Comment (st4rl3ss) [reply var/spack/repos/builtin/packages/neurodamus-asciitoh/package.py]: I've just copied the relevant files to the `bin` directory. For everything else I kept the `neurodamus` structure as Iannis suggested.

File: var/spack/repos/builtin/packages/neurodamus-asciitoh/package.py (hunk 4 above)
Comment (pramodk) [reply var/spack/repos/builtin/packages/neurodamus-asciitoh/package.py]: > If you think python is useless I can remove it.
I would try `neuron~mpi`. If builds, there is no reason to add more constraints.

File: var/spack/repos/builtin/packages/neurodamus-asciitoh/package.py (hunk 5 above)
Comment (pramodk) [reply var/spack/repos/builtin/packages/neurodamus-asciitoh/package.py]: that's good unless we are duplicating the same function in this package. I think we need is :

```
spec['hdf5'].libs.ld_flags
spec['hdf5'].libs.rpath_flags
```
You can check by printing above flags and may be just add to link_flag ?

File: var/spack/repos/builtin/packages/neurodamus-asciitoh/package.py (hunk 2 above)
Comment (st4rl3ss) [reply var/spack/repos/builtin/packages/neurodamus-asciitoh/package.py]: This is the issue I'm having: 
```
 spack create neurodamus-asciitoh5     
==> This looks like a URL for neurodamus-asciitoh
==> Warning: Spack will not check SSL certificates. You need to update your Python to enable certificate verification.
==> Warning: Spack will not check SSL certificates. You need to update your Python to enable certificate verification.
==> Found 1 version of neurodamus-asciitoh:
  
  5  neurodamus-asciitoh5

==> How many would you like to checksum? (default is 1, q to abort) 1
==> Downloading...
==> Fetching neurodamus-asciitoh5

curl: (6) Could not resolve host: neurodamus-asciitoh5; Name or service not known
==> Fetching from neurodamus-asciitoh5 failed.
==> Something failed on neurodamus-asciitoh5, skipping.
    (All fetchers failed for spack-stage-SlmBor)
==> Error: Could not fetch any versions for neurodamus-asciitoh
```

Comment (pramodk) [reply var/spack/repos/builtin/packages/neurodamus-asciitoh/package.py]:
```
curl: (6) Could not resolve host: neurodamus-asciitoh5
```
as error says, `curl` command  is trying to download from web server name `neurodamus-asciitoh5`. (which is wrong). Lets check offline.

File: var/spack/repos/builtin/packages/asciitoh5/package.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+class Asciitoh5(Package):
+    """Neurodamus Library necessary to convert from ASCII to H5"""
+   
+    homepage = "ssh://bbpcode.epfl.ch/sim/utils/asciitoh5"
+    git      = "ssh://bbpcode.epfl.ch/sim/utils/asciitoh5"
+
+    version('develop', git=git, branch='master')
+    version('0.4', git=git, tag='0.4')
+    version('0.2', git=git, tag='0.2')
+    version('0.1', git=git, tag='0.1')
```
Comment (pramodk) [inline var/spack/repos/builtin/packages/asciitoh5/package.py]: we can remove 0.2 and 0.1 if those were not useful?

**Review by pramodk — COMMENTED** [review]
LGTM. we can check local installation quickly on your system.

File: var/spack/repos/builtin/packages/asciitoh5/package.py (hunk 7 above)
Comment (st4rl3ss) [reply var/spack/repos/builtin/packages/asciitoh5/package.py]: Right! Removed.

**Comment by st4rl3ss** [comment]
Seems to be working fine now. Also tested by sourcing:
```
/gpfs/bbp.cscs.ch/apps/hpc/jenkins/pulls/465/config/modules.sh
```

**Review by matz-e — APPROVED** [review]

**Review by pramodk — APPROVED** [review]
LGTM. Please squash and merge.

Could you verify the generated module in the pulls directory on bb5 is working?

**Comment by st4rl3ss** [comment]
I had already tested that past week. I'll squash and merge!

**Pull request merged**